### PR TITLE
Expose compaction metrics to Prometheus

### PR DESCRIPTION
**Pull request by Technoboy-**
### Motivation
As #11564 has involved compaction metrics in CLI, it's extremely useful to expose relative metrics to Prometheus.

- pulsar_compaction_removed_event_count :  the removed event count of compaction .
- pulsar_compaction_succeed_count : the succeed count of compaction.
- pulsar_compaction_failed_count : the failed count of compaction.
- pulsar_compaction_duration_time_in_millis : the duration time of compaction.
- pulsar_compaction_read_throughput : the read throughput of compaction.
- pulsar_compaction_write_throughput : the write throughput of compaction.
- pulsar_compaction_compacted_entries_count: the compacted entries count.
- pulsar_compaction_compacted_entries_size: the compacted entries size;

if users enable the topic level metrics and the topic has been compacted or doing the compaction, we should expose the compaction metrics for the topic level. If users disable the topic level metrics, we should expose the aggregated compaction metrics for the namespace level.

#### For contributor

The document is required.

### CODE REVIEW

**Comment by Anonymitaet** [comment]
Thanks for your contribution. Please do not forget to update docs later. And you can ping me to review the docs, thanks.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 1]
```diff
@@ -2120,6 +2115,15 @@ public TopicStatsImpl getStats(boolean getPreciseBacklog, boolean subscriptionBa
         return statFuture;
     }
 
+    public Optional<CompactedTopicImpl.CompactedTopicContext> getCompactedTopicContext() {
```
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:2118]: exposing something named *Impl is a code smell.

can we create an interface for `CompactedTopicContext` ?

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/stats/CompactionMetricsTest.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            byte[] data = ("my-message-" + key + "-" + j).getBytes();
+            producer.newMessage()
+                    .key(key)
+                    .value(data)
+                    .send();
+        }
+
+        pulsar.getCompactor(true).compact(topic).get();
+        List<Metrics> list2 = metrics.generate();
+        System.out.println(list2);
```
Comment (eolivelli) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/stats/CompactionMetricsTest.java:88]: nit: please remove System.out or use logger

**Review by eolivelli — CHANGES_REQUESTED** [review]
very good.

I left one comment about exposing an "Impl" class in PersistentTopic.

PTAL

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 1 above)
Comment (Technoboy-) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java]: Yes.  I have made  CompactedTopicContext as a public class.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/stats/CompactionMetricsTest.java (hunk 2 above)
Comment (Technoboy-) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/stats/CompactionMetricsTest.java]: Done.

**Comment by Technoboy-** [comment]
> very good.
> 
> I left one comment about exposing an "Impl" class in PersistentTopic.
> 
> PTAL

Thanks for reviewing.

**Review by eolivelli — APPROVED** [review]
LGTM

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/metrics/AbstractMetrics.java [hunk 3]
```diff
@@ -242,24 +238,11 @@ protected void populateMaxMap(Map<String, Long> map, String mkey, long value) {
...
             Metrics metrics, TopicStats destStats) {
-        if (!topicsStatsByDimensionMap.containsKey(metrics)) {
-            // create new list
-            topicsStatsByDimensionMap.put(metrics, Lists.newArrayList(destStats));
-        } else {
-            // add to collection
-            topicsStatsByDimensionMap.get(metrics).add(destStats);
-        }
-
+        topicsStatsByDimensionMap.computeIfAbsent(metrics, __ -> Lists.newArrayList(destStats)).add(destStats);
```
[the remark this answers is not in the document]
Comment (Technoboy-) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/metrics/AbstractMetrics.java]: Done.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/metrics/CompactionMetrics.java [hunk 4]
```diff
@@ -159,10 +161,6 @@ public CompactionMetrics(PulsarService pulsar) {
     }
 
     private void populateDimensionMap(Map<Metrics, List<String>> topicsByDimensionMap, Metrics metrics, String topic) {
-        if (!topicsByDimensionMap.containsKey(metrics)) {
-            topicsByDimensionMap.put(metrics, Lists.newArrayList(topic));
-        } else {
-            topicsByDimensionMap.get(metrics).add(topic);
-        }
+        topicsByDimensionMap.computeIfAbsent(metrics, __ -> Lists.newArrayList(topic)).add(topic);
```
[the remark this answers is not in the document]
Comment (Technoboy-) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/metrics/CompactionMetrics.java]: Yes, thanks.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/metrics/AbstractMetrics.java [hunk 5]
```diff
@@ -242,24 +238,11 @@ protected void populateMaxMap(Map<String, Long> map, String mkey, long value) {
...
     protected void populateDimensionMap(Map<Metrics, List<ManagedLedgerImpl>> ledgersByDimensionMap, Metrics metrics,
             ManagedLedgerImpl ledger) {
-        if (!ledgersByDimensionMap.containsKey(metrics)) {
-            // create new list
-            ledgersByDimensionMap.put(metrics, Lists.newArrayList(ledger));
-        } else {
-            // add to collection
-            ledgersByDimensionMap.get(metrics).add(ledger);
-        }
+        ledgersByDimensionMap.computeIfAbsent(metrics, __ -> Lists.newArrayList(ledger)).add(ledger);
```
[the remark this answers is not in the document]
Comment (Technoboy-) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/metrics/AbstractMetrics.java]: Done

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/metrics/AbstractMetrics.java [hunk 6]
```diff
@@ -214,11 +214,7 @@ protected Metrics createMetricsByDimension(String namespace, String fromClusterN
     }
 
     protected void populateAggregationMap(Map<String, List<Double>> map, String mkey, double value) {
-        if (!map.containsKey(mkey)) {
-            map.put(mkey, Lists.newArrayList(value));
-        } else {
-            map.get(mkey).add(value);
-        }
+        map.computeIfAbsent(mkey, __ -> Lists.newArrayList(value)).add(value);
```
[the remark this answers is not in the document]
Comment (Technoboy-) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/metrics/AbstractMetrics.java]: Done

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java [hunk 7]
```diff
@@ -1999,6 +1999,11 @@
             doc = "If true, export managed cursor metrics"
     )
     private boolean exposeManagedCursorMetricsInPrometheus = false;
+    @FieldContext(
+            category = CATEGORY_METRICS,
+            doc = "If true, export compaction metrics"
+    )
+    private boolean exposeCompactionMetricsInPrometheus = false;
```
Comment (codelipenghui) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java:2006]: In my opinion, we don't need this new flag to enable or disable the `feature` metrics.
The existing metrics control flag is based on the different levels:

```
# Enable topic level metrics
exposeTopicLevelMetricsInPrometheus=true

# Enable consumer level metrics. default is false
exposeConsumerLevelMetricsInPrometheus=false

# Enable producer level metrics. default is false
exposeProducerLevelMetricsInPrometheus=false

# Enable managed ledger metrics (aggregated by namespace). default is false
exposeManagedLedgerMetricsInPrometheus=true

# Enable cursor level metrics. default is false
exposeManagedCursorMetricsInPrometheus=false
```

If we introduce a feature-based metrics control, will make the configuration more complex such as enable `feature` metrics on the level.

My suggestion is if users enable the topic level metrics and the topic has been compacted or doing the compaction, we should expose the compaction metrics for the topic level. If users disable the topic level metrics, we should expose the aggregated compaction metrics for the namespace level.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 8]
```diff
@@ -1906,14 +1908,19 @@ public TopicStatsImpl getStats(boolean getPreciseBacklog, boolean subscriptionBa
...
-                stat.getLastCompactionSucceedTimestamp(topic)).orElse(0L);
-        stats.compaction.lastCompactionFailedTimestamp = mxBean.map(stat ->
-                stat.getLastCompactionFailedTimestamp(topic)).orElse(0L);
-        stats.compaction.lastCompactionDurationTimeInMills = mxBean.map(stat ->
-                stat.getLastCompactionDurationTimeInMills(topic)).orElse(0L);
+
+        stats.compaction.lastCompactionRemovedEventCount = 0L;
+        stats.compaction.lastCompactionSucceedTimestamp = 0L;
+        stats.compaction.lastCompactionFailedTimestamp = 0L;
+        stats.compaction.lastCompactionDurationTimeInMills = 0L;
```
Comment (hangc0276) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:1915]:
```java
stats.compaction.lastCompactionRemovedEventCount = 0L;
stats.compaction.lastCompactionSucceedTimestamp = 0L;
stats.compaction.lastCompactionFailedTimestamp = 0L;
stats.compaction.lastCompactionDurationTimeInMills = 0L;
```
replace with  `stats.compaction.reset(); ` ?

**Review by hangc0276 — COMMENTED** [review]
Please add a test to check the final prometheus metric output result in `PrometheusMetricsTest`

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java (hunk 7 above)
Comment (Technoboy-) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java]: Ok, updated.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 8 above)
Comment (Technoboy-) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:1915]: Done

**Review by michaeljmarshall — APPROVED** [review]
LGTM. Thanks for addressing all of my feedback.

**Comment by Technoboy-** [comment]
> LGTM. Thanks for addressing all of my feedback.

Many thanks for your kind review.

**Comment by Technoboy-** [comment]
> Please add a test to check the final prometheus metric output result in `PrometheusMetricsTest`

Yes, done. Thanks for reviewing.

**Review by codelipenghui — APPROVED** [review]

**Comment by Anonymitaet** [comment]
Hi @sijia-w, in this PR, @Technoboy- has already added docs to`master`. 

- Could you please add the same contents to `2.8.2`?
- If you find any inaccurate expressions or grammar issues, you can modify the docs based on technical writing rules.

Thanks!

**Pull request merged**